index.js

```diff
@@ -0,0 +1,693 @@
+        var typeClasses = constraints[t.name];
+        for (idx = 0; idx < typeClasses.length; idx += 1) {
+          if (!typeClasses[idx].test(value)) {
+            throw new TypeError(
+              LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
+              ' requires ' + LEFT_SINGLE_QUOTATION_MARK + t.name +
+              RIGHT_SINGLE_QUOTATION_MARK + ' to implement ' +
+              constraints[t.name].join(' and ') + '; ' +
+              actualTypes.join(' and ') + ' ' +
+              (actualTypes.length === 1 ? 'does' : 'do') + ' not'
+            );
+          }
+        }
+      }
+
+      if (t.type === 'TypeVariable' && !has(t.name, typeVarMap)) {
+        var consistentTypes = filter(actualTypes, function(t) { return t.type !== 'InconsistentType'; });
+        if (consistentTypes.length === 0) {
+          return typeVarMap;
+        } else {
+          var $typeVarMap = {};
+          $typeVarMap[t.name] = consistentTypes;
+          var names = keys(typeVarMap);
+          for (idx = 0; idx < names.length; idx += 1) {
+            $typeVarMap[names[idx]] = typeVarMap[names[idx]];
+          }
+          return $typeVarMap;
+        }
+      } else if (t.type === 'UnaryType') {
+        for (idx = 0; idx < actualTypes.length; idx += 1) {
+          actualType = actualTypes[idx];
+          if (actualType.type === t.type &&
+              actualType.name === t.name &&
+              equalTypes(actualType.$1, t.$1)) {
+            typeVarMap = reduce(
+              actualType._1(value),
+              typeVarMap,
+              function(typeVarMap, innerValue) {
+                return recur(name, constraints, typeVarMap,
+                             [t.$1], [actualType.$1], innerValue);
+              }
+            );
+          }
+        }
+        return typeVarMap;
+      } else if (t.type === 'BinaryType') {
+        for (idx = 0; idx < actualTypes.length; idx += 1) {
+          actualType = actualTypes[idx];
+          if (actualType.type === t.type &&
+              actualType.name === t.name &&
+              equalTypes(actualType.$1, t.$1) &&
+              equalTypes(actualType.$2, t.$2)) {
+            typeVarMap = reduce(
+              actualType._1(value),
+              typeVarMap,
+              function(typeVarMap, innerValue) {
+                return recur(name, constraints, typeVarMap,
+                             [t.$1], [actualType.$1], innerValue);
+              }
+            );
+            typeVarMap = reduce(
+              actualType._2(value),
+              typeVarMap,
+              function(typeVarMap, innerValue) {
+                return recur(name, constraints, typeVarMap,
+                             [t.$2], [actualType.$2], innerValue);
+              }
+            );
+          }
+        }
+        return typeVarMap;
+      } else {
+        return typeVarMap;
+      }
+    });
+  };
+
+  var doAllTheThings = function(o) {
```

Comment:
It was a sign that I'd lost control of the code. The new implementation does not contain a function named `doAllTheThings`. :)